### PR TITLE
Fixing when name is empty, support clientcredentialtype and refractoring

### PR DESCRIPTION
**Pull request by birojnayak**
1)Support for ClientCredential type across all bindings
2)Fixing when Name is empty #445 
3)Refactoring and adding test

### CODE REVIEW

**Review by Eruanion — APPROVED** [review]

File: src/CoreWCF.ConfigurationManager/src/CoreWCF/Configuration/HttpTransportSecurityElement.cs [hunk 1]
```diff
@@ -44,5 +44,18 @@ public string Realm
...
+        internal void ApplyConfiguration(HttpTransportSecurity security)
+        {
+            if (security == null)
+            {
+                throw DiagnosticUtility.ExceptionUtility.ThrowHelperArgumentNull(nameof(security));
+            }
+
+            security.ClientCredentialType = ClientCredentialType;
+           // security.ProxyCredentialType = this.ProxyCredentialType;
+            security.Realm = Realm;
```
Comment (mconnew) [inline src/CoreWCF.ConfigurationManager/src/CoreWCF/Configuration/HttpTransportSecurityElement.cs:57]: ProxyCredentialType is a client side property only so it's not a case of we haven't implemented this yet, we will never add that property. It's safe to remove this. Leave ExtendedProtectionPolixy though, we will need to add that eventually

File: src/CoreWCF.ConfigurationManager/src/CoreWCF/Configuration/NetHttpBindingElement.cs [hunk 2]
```diff
@@ -65,6 +65,10 @@ public override Binding CreateBinding()
                 ReaderQuotas = ReaderQuotas.Clone(),
             };
 
+            binding.MessageEncoding = MessageEncoding;
+            //WebSocketSettings.ApplyConfiguration(netHttpBinding.WebSocketSettings);
+            // this.ReliableSession.ApplyConfiguration(netHttpBinding.ReliableSession);
```
Comment (mconnew) [inline src/CoreWCF.ConfigurationManager/src/CoreWCF/Configuration/NetHttpBindingElement.cs:70]: Can you open a general purpose issue for adding more capabilities to ConfigurationManager with a list with check boxes in the main comment (added by adding square brackets at start of line \[\]) and add the currently known missing features. The one's I can think of off the top of my head are behaviors, base addresses, and now WebSockets for NetHttpBinding. As they get fixed, will mark them as done.

File: src/CoreWCF.ConfigurationManager/src/CoreWCF/Configuration/NetTcpBindingElement.cs [hunk 3]
```diff
@@ -126,6 +126,9 @@ public override Binding CreateBinding()
                 HostNameComparisonMode = HostNameComparisonMode
             };
 
+            //this.ReliableSession.ApplyConfiguration(nptBinding.ReliableSession);
+            Security.ApplyConfiguration(binding.Security);
+            //this.ReaderQuotas.ApplyConfiguration(nptBinding.ReaderQuotas);
             return binding;
```
Comment (mconnew) [inline src/CoreWCF.ConfigurationManager/src/CoreWCF/Configuration/NetTcpBindingElement.cs:132]: The ReaderQuotas are being applied on line 124 so you can remove this line. Although XmlDictionaryReaderQuotasElement.Clone() should really be called something like CreateReaderQuotas() as it's not really cloning. Up to you if you want to fix that naming, it's internal anyway.

File: src/CoreWCF.ConfigurationManager/src/CoreWCF/Configuration/ServiceReflector.cs [hunk 4]
```diff
@@ -18,7 +18,7 @@ internal static Type ResolveTypeFromName(string name)
                 }
             }
 
-            return default;
+            return Type.GetType(name, true);
         }
```
Comment (mconnew) [inline src/CoreWCF.ConfigurationManager/src/CoreWCF/Configuration/ServiceReflector.cs:22]: Could you add a comment at the top along the lines of checking if type is in an already loaded assembly first and then before this line add a comment about the type string needing to be an AssemblyQualifiedName to make CLR load the assembly.

File: src/CoreWCF.ConfigurationManager/src/CoreWCF/Configuration/TcpTransportSecurityElement.cs [hunk 5]
```diff
@@ -36,5 +36,18 @@ public TcpClientCredentialType ClientCredentialType
         //    get { return (SslProtocols)base[ConfigurationStrings.SslProtocols]; }
         //    private set { base[ConfigurationStrings.SslProtocols] = value; }
         //}
+
+        internal void ApplyConfiguration(TcpTransportSecurity security)
+        {
```
Comment (mconnew) [inline src/CoreWCF.ConfigurationManager/src/CoreWCF/Configuration/TcpTransportSecurityElement.cs:41]: We do have support for SslProtocols in the implementation. Can you uncomment those pieces?

File: src/CoreWCF.ConfigurationManager/src/CoreWCF/Configuration/WSHttpTransportSecurityElement.cs [hunk 6]
```diff
@@ -43,5 +43,18 @@ public string Realm
...
+        internal void ApplyConfiguration(HttpTransportSecurity security)
+        {
+            if (security == null)
+            {
+                throw DiagnosticUtility.ExceptionUtility.ThrowHelperArgumentNull("security");
+            }
+
+            security.ClientCredentialType = ClientCredentialType;
+           // security.ProxyCredentialType = this.ProxyCredentialType;
+            security.Realm = Realm;
```
Comment (mconnew) [inline src/CoreWCF.ConfigurationManager/src/CoreWCF/Configuration/WSHttpTransportSecurityElement.cs:56]: Same as earlier comment, proxy is client side only so remove.

File: src/CoreWCF.ConfigurationManager/src/CoreWCF/Configuration/WSHttpTransportSecurityElement.cs [hunk 7]
```diff
@@ -43,5 +43,18 @@ public string Realm
                 base[ConfigurationStrings.Realm] = value;
             }
         }
+
+        internal void ApplyConfiguration(HttpTransportSecurity security)
+        {
+            if (security == null)
+            {
+                throw DiagnosticUtility.ExceptionUtility.ThrowHelperArgumentNull("security");
+            }
```
Comment (mconnew) [inline src/CoreWCF.ConfigurationManager/src/CoreWCF/Configuration/WSHttpTransportSecurityElement.cs:52]: nameof

File: src/CoreWCF.ConfigurationManager/tests/BasicHttpBindingTests.cs [hunk 8]
```diff
@@ -31,7 +32,9 @@ public void BasicHttpBinding_WithCustomSetting()
                          maxReceivedMessageSize=""{expectedMaxReceivedMessageSize}""
                          maxBufferSize=""{expectedMaxBufferSize}""
                          receiveTimeout=""00:10:00"">
-                    <security mode=""{expectedSecurityMode}""/>
+                    <security mode=""{expectedSecurityMode}"">
+                     <message clientCredentialType=""Certificate"" />
+                     </security>
```
Comment (mconnew) [inline src/CoreWCF.ConfigurationManager/tests/BasicHttpBindingTests.cs:37]: Should this use `{clientCredType}` like the other properties?

File: src/CoreWCF.ConfigurationManager/tests/WSHttpBindingTests.cs [hunk 9]
```diff
@@ -30,7 +31,9 @@ public void WSHttpBinding_WithCustomSetting()
                          maxReceivedMessageSize=""{expectedMaxReceivedMessageSize}""
                          maxBufferPoolSize=""{expectedMaxBufferPoolSize}""
                          receiveTimeout=""00:10:00"">
-                    <security mode=""{expectedSecurityMode}""/>
+                    <security mode=""{expectedSecurityMode}"">
+                    <message clientCredentialType=""UserName"" />
+                     </security>
```
Comment (mconnew) [inline src/CoreWCF.ConfigurationManager/tests/WSHttpBindingTests.cs:36]: You didn't add an Assert below to make sure this is read correctly.

File: src/CoreWCF.ConfigurationManager/tests/WSHttpBindingTests.cs [hunk 10]
```diff
@@ -61,6 +64,13 @@ public void WSHttpBinding_WithCustomSetting()
         [Fact]
         public void WSHttpBinding_WithDefaultSetting()
         {
+            //TODO:- Check if there is a better way to skip the test
+            string frameworkDescription = RuntimeInformation.FrameworkDescription;
+            if (frameworkDescription.IndexOf(".NET Framework", StringComparison.Ordinal) >= 0)
```
Comment (mconnew) [inline src/CoreWCF.ConfigurationManager/tests/WSHttpBindingTests.cs:69]: Have a look at Trait's. I added one for some tests to only be run on Windows. You can add a trait for NetCore and add that to the pipeline yml file for test runs on non-NetFx platforms.

**Review by mconnew — CHANGES_REQUESTED** [review]
:clock1:

File: src/CoreWCF.ConfigurationManager/src/CoreWCF/Configuration/NetHttpBindingElement.cs (hunk 2 above)
Comment (birojnayak) [reply src/CoreWCF.ConfigurationManager/src/CoreWCF/Configuration/NetHttpBindingElement.cs:70]: https://github.com/CoreWCF/CoreWCF/issues/446 => I will sync offline to modify

File: templates/TestStage.yml [hunk 11]
```diff
@@ -99,7 +99,7 @@ stages:
         Windows_netfx:
           imageName: 'windows-latest'
           targetFramework: 'net472'
-          testArgs: ''
+          testArgs: '--filter Category!=NetCoreOnly'
         Linux_netcore2.1:
```
Comment (mconnew) [inline templates/TestStage.yml:103]: You need to add this to TestRelease as well.

File: src/CoreWCF.ConfigurationManager/src/CoreWCF/Configuration/WSHttpTransportSecurityElement.cs [hunk 12]
```diff
@@ -43,5 +43,17 @@ public string Realm
...
+        internal void ApplyConfiguration(HttpTransportSecurity security)
+        {
+            if (security == null)
+            {
+                throw DiagnosticUtility.ExceptionUtility.ThrowHelperArgumentNull(nameof(security));
+            }
+
+            security.ClientCredentialType = ClientCredentialType;
+            security.Realm = Realm;
+           // security.ExtendedProtectionPolicy = ChannelBindingUtility.BuildPolicy(this.ExtendedProtectionPolicy);
```
Comment (mconnew) [inline src/CoreWCF.ConfigurationManager/src/CoreWCF/Configuration/WSHttpTransportSecurityElement.cs:56]: You forgot to remove the commented out property for ProxyCredentialType above when you removed the commented setting code here.

**Review by mconnew — APPROVED** [review]

**Pull request merged**